### PR TITLE
feat(Validium): `PubdataPricing` trait

### PR DESCRIPTION
**Pull request by ColoCarletti**
## What ❔

Implement `PubdataPricing` trait, with  `ValidiumPubdataPricing` and `RollupPubdataPricing` to use it in `estimate_effective_pubdata_price()`

## Why ❔

It was agreed in this comment https://github.com/matter-labs/zksync-era/pull/1238#discussion_r1502087828,

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.
- [x] Spellcheck has been run via `zk spellcheck`.
- [x] Linkcheck has been run via `zk linkcheck`.

### CODE REVIEW

File: core/lib/zksync_core/src/l1_gas_price/gas_adjuster/tests.rs [hunk 1]
```diff
@@ -27,8 +28,7 @@ fn samples_queue() {
 }
 
 /// Check that we properly fetch base fees as block are mined
-#[tokio::test]
-async fn kept_updated() {
+async fn _keep_updated(pubdata_pricin: Arc<dyn PubdataPricing>) {
```
Comment (ilitteri) [inline core/lib/zksync_core/src/l1_gas_price/gas_adjuster/tests.rs:31]: There's a typo here

File: core/lib/zksync_core/src/state_keeper/io/tests/mod.rs [hunk 2]
```diff
@@ -26,119 +24,47 @@ use crate::{
...
 
-    // Genesis is needed for proper mempool initialization.
-    tester.genesis(&connection_pool).await;
-    let (mempool, _) = tester.create_test_mempool_io(connection_pool, 1).await;
-
-    // Upon initialization, the filter should be set to the default values.
-    assert_eq!(mempool.filter(), &L2TxFilter::default());
+    tests_helpers::test_filter_initialization(rollup_tester).await;
+    tests_helpers::test_filter_initialization(validium_tester).await;
 }
```
Comment (popzxc) [inline core/lib/zksync_core/src/state_keeper/io/tests/mod.rs:40]: We already depend on [test_casing](https://docs.rs/test-casing/latest/test_casing/) crate, so I think you can just reuse it for testing both Rollup/Validium.
You can see the sample usage [here](https://github.com/matter-labs/zksync-era/blob/main/core/lib/zksync_core/src/sync_layer/batch_status_updater/tests.rs#L344).

**Review by popzxc — COMMENTED** [review]
Mostly LGTM, except for tests.

File: core/lib/zksync_core/src/state_keeper/io/tests/mod.rs [hunk 3]
```diff
@@ -656,11 +673,11 @@ async fn miniblock_sealer_handle_parallel_processing() {
     sealer_handle.wait_for_all_commands().await;
 }
 
-/// Ensure that subsequent miniblocks that belong to the same L1 batch have different timestamps
```
Comment (popzxc) [inline core/lib/zksync_core/src/state_keeper/io/tests/mod.rs:659]: I think this doc-comment was removed by mistake.

**Review by popzxc — APPROVED** [review]
Nice.

File: core/lib/zksync_core/src/state_keeper/io/tests/mod.rs (hunk 2 above)
Comment (popzxc) [reply core/lib/zksync_core/src/state_keeper/io/tests/mod.rs]: You mean test modules for test-casing purposes? It's fine, I think. The types are cheap and on the first glance it looks better than introducing some kind of shared knowledge between purely test code.

**Comment by ilitteri** [comment]
The integration tests that are failing are fixed in https://github.com/matter-labs/zksync-era/pull/1270 and https://github.com/matter-labs/zksync-era/pull/1296

**Pull request merged**